odoo/fields.py

```diff
@@ -894,10 +894,17 @@ def update_db_notnull(self, model, column):
                 # flush values before adding NOT NULL constraint
                 model.flush([self.name])
 
-        if self.required and not has_notnull:
-            model.pool.post_constraint(sql.set_not_null, model._cr, model._table, self.name)
-        elif not self.required and has_notnull:
-            sql.drop_not_null(model._cr, model._table, self.name)
+        if self.required:
+            if not has_notnull:
+                fail = model.pool.post_constraint(sql.set_not_null, model._cr, model._table, self.name)
```

Comment:
I would rather use the constraint handling to our advantage here:
```suggestion
                model.pool.post_constraint(apply_required, model._cr, self.model_name, self.name)
```
where the function `apply_required` would look like
```python
import odoo

def apply_required(cr, model_name, field_name):
    """ Set the NOT NULL constraint on the given field, if necessary. """
    # retrieve the field
    registry = odoo.registry(cr.dbname)
    Model = registry[model_name]
    field = Model._fields[field_name]

    # set the NOT NULL constraint if it still applies
    if field.store and field.required:
        # todo: catch potential error if the constraint already exists
        sql.set_not_null(cr, Model._table, field_name)
```
Note that the registry or the model could be passed as a parameter to the function, but one has to make sure that the registry has not been reloaded in the mean time, and the model's class is still the same.  However, the field object cannot be passed as a parameter, since the model setup recreates that object.

The nice part is that **no extra stuff is needed anywhere else:**
- if the first attempt fails, the failure is caught by the registry;
- if the second attempt fails, an error message is shown;
- **if the field is no longer required at the second attempt, the function does nothing**.

Reply:
Apparently, setting the constraint NOT NULL on an already NOT NULL columns gracefully succeeds.  So one can apply the NOT NULL constraint twice without problem.